File: backend/app/schemas/notification.py
from datetime import datetime
from typing import List, Literal, Optional
from pydantic import BaseModel, ConfigDict, Field

NotificationType = Literal["info", "success", "warning", "danger"]
NotificationCategory = Literal["attendance", "leave", "device", "security", "system"]


class NotificationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    user_id: Optional[str] = Field(None, description="Target user ID if user-specific")
    role: Optional[str] = Field(None, description="Role target if broadcast")
    title: str
    message: str
    type: NotificationType = "info"
    category: NotificationCategory = "system"
    link: Optional[str] = None
    is_read: bool = Field(False, description="Read status for the notification")
    created_at: datetime


class NotificationListResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    items: List[NotificationResponse]
    total_count: int
    unread_count: int


class NotificationBroadcastCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str = Field(..., min_length=2, max_length=150)
    message: str = Field(..., min_length=2, max_length=1000)
    target_role: Optional[Literal["ALL", "TEACHER", "STUDENT", "ADMIN"]] = "ALL"
    type: NotificationType = "info"
    category: NotificationCategory = "system"
    link: Optional[str] = Field(None, max_length=255)

File: frontend/src/types/notification.ts
export type NotificationType = "info" | "success" | "warning" | "danger";

export type NotificationCategory =
  | "attendance"
  | "leave"
  | "device"
  | "security"
  | "system";

export interface NotificationItem {
  id: string;
  user_id?: string | null;
  role?: string | null;
  title: string;
  message: string;
  type: NotificationType;
  category: NotificationCategory;
  link?: string | null;
  is_read: boolean;
  created_at: string;
}

export interface NotificationListResponse {
  items: NotificationItem[];
  total_count: number;
  unread_count: number;
}

export interface NotificationBroadcastCreate {
  title: string;
  message: string;
  target_role?: "ALL" | "TEACHER" | "STUDENT" | "ADMIN";
  type?: NotificationType;
  category?: NotificationCategory;
  link?: string | null;
}

export type NotificationFilterTab = "all" | "unread" | "attendance" | "system";
